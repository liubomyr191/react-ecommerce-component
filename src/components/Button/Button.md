A typical example:

```jsx
<Button title="Submit" className="myBtn">Submit</Button>
```
