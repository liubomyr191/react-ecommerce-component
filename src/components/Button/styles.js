export const containerStyles = {
  cursor: "pointer",
  marginBottom: 0,
  marginLeft: 0,
  marginRight: 0,
  marginTop: 0,
  paddingBottom: "0.5rem",
  paddingLeft: "1.5rem",
  paddingRight: "1.5rem",
  paddingTop: "0.5rem",
  textAlign: "center"
};

export const flexContainerStyles = {
  ...containerStyles,
  display: "flex"
};
