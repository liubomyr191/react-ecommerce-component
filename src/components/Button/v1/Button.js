import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { applyTheme, preventAccidentalDoubleClick } from "helpers";
import { applyThemeWithActionType } from "./helpers";

const paddingFunc = (props) => {
  const { isShortHeight } = props;
  if (isShortHeight) return applyTheme("buttonVerticalPaddingShort");
  return applyTheme("buttonVerticalPadding");
};

const ButtonDiv = styled.div`
  background-color: ${applyThemeWithActionType("buttonBackgroundColor", true)};
  border-color: ${applyThemeWithActionType("buttonBorderColor", true)};
  border-style: solid;
  border-width: 1px;
  border-radius: ${applyTheme("buttonBorderRadius")};
  box-sizing: border-box;
  color: ${applyThemeWithActionType("buttonForegroundColor", false)};
  cursor: pointer;
  display: ${(props) => {
    const { children, fullWidth } = props;
    if (fullWidth) {
      if (typeof children === "string") return "block";
      return "flex";
    }
    if (typeof children === "string") return "inline-block";
    return "inline-flex";
  }};
  margin: 0;
  min-width: ${applyTheme("buttonMinimumWidth")};
  outline: none;
  padding-left: ${applyTheme("buttonHorizontalPadding")};
  padding-right: ${applyTheme("buttonHorizontalPadding")};
  padding-top: ${paddingFunc};
  padding-bottom: ${paddingFunc};
  text-align: center;

  &:hover {
    background-color: ${applyThemeWithActionType("buttonBackgroundColor", true, "hover")};
    border-color: ${applyThemeWithActionType("buttonBorderColor", true, "hover")};
    color: ${applyThemeWithActionType("buttonForegroundColor", false, "hover")};
  }

  &:active {
    background-color: ${applyThemeWithActionType("buttonBackgroundColor", true, "active")};
    border-color: ${applyThemeWithActionType("buttonBorderColor", true, "active")};
    color: ${applyThemeWithActionType("buttonForegroundColor", false, "active")};
  }
`;

class Button extends Component {
  static propTypes = {
    /**
     * The type of action performed by the button
     */
    actionType: PropTypes.oneOf(["danger", "default", "important", "secondary", "secondaryDanger"]),
    /**
     * The contents of the button, such as text, icons, or any combination of React and HTML components
     */
    children: PropTypes.node,
    /**
     * Classes for the button element
     */
    className: PropTypes.string,
    /**
     * Set to `true` to prevent the button from calling `onClick` when clicked
     */
    isDisabled: PropTypes.bool,
    /**
     * Button should take full width
     */
    isFullWidth: PropTypes.bool,
    /**
     * Enable this when you donâ€™t have enough vertical space, such as in table headers
     */
    isShortHeight: PropTypes.bool,
    /**
     * Enable this in rare cases where having a solid or outline button for an action would be cluttered and visually confusing.
     */
    isTextOnly: PropTypes.bool,
    /**
     * Called with no arguments whenever the button is clicked. There is double-click protection,
     * so if the user double-clicks quickly, onClick is called only once.
     */
    onClick: PropTypes.func,
    /**
     * The string for the `title` attribute of the button element
     */
    title: PropTypes.string
  };

  static defaultProps = {
    actionType: "default",
    isDisabled: false,
    isFullWidth: false,
    onClick() {}
  };

  handleClick = preventAccidentalDoubleClick((event) => {
    event.preventDefault();

    const { isDisabled, onClick } = this.props;

    if (!isDisabled) onClick();
  });

  handleKeyPress = (event) => {
    if (event.keyCode === 13) this.handleClick(event);
  };

  render() {
    const { actionType, children, className, isDisabled, isFullWidth, isShortHeight, isTextOnly, title } = this.props;

    const moreButtonDivProps = {};
    if (isDisabled) {
      moreButtonDivProps["aria-disabled"] = "true";
    }

    return (
      <ButtonDiv
        actionType={actionType}
        className={className}
        isDisabled={isDisabled}
        fullWidth={isFullWidth}
        onClick={this.handleClick}
        onKeyPress={this.handleKeyPress}
        role="button"
        isShortHeight={isShortHeight}
        isTextOnly={isTextOnly}
        tabIndex={0}
        title={title}
        {...moreButtonDivProps}
      >
        {children}
      </ButtonDiv>
    );
  }
}

export default Button;
