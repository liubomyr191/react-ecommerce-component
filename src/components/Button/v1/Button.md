The `Button` component can render with four different possible appearances, which are chosen based on the `actionType` prop value. You should choose the appropriate appearance based on what type of action will happen if you click it.

The actual colors of the appearance are defined by your theme. The examples here use our default theme. The theme properties used by this component are:

- `rui_buttonBorderRadius`
- `rui_buttonBorderColor_danger`
- `rui_buttonBorderColor_default`
- `rui_buttonBorderColor_important`
- `rui_buttonBorderColor_secondary`
- `rui_buttonBackgroundColor_danger`
- `rui_buttonBackgroundColor_default`
- `rui_buttonBackgroundColor_important`
- `rui_buttonBackgroundColor_secondary`
- `rui_buttonForegroundColor_danger`
- `rui_buttonForegroundColor_default`
- `rui_buttonForegroundColor_important`
- `rui_buttonForegroundColor_secondary`
- `rui_buttonPadding`

#### Default Button

The default button is the most common button and should be your first choice.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Default" className="myBtn">Default</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Default" className="myBtn" disabled>Default Disabled</Button>
  </div>
</div>
```

#### Secondary Button

For secondary actions or actions that are not required. For example, a Cancel button.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Secondary" className="myBtn" actionType="secondary">Secondary</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Secondary" className="myBtn" actionType="secondary" disabled>Secondary Disabled</Button>
  </div>
</div>
```

#### Important Button

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Important" className="myBtn" actionType="important">Important</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Important" className="myBtn" actionType="important" disabled>Important Disabled</Button>
  </div>
</div>
```

#### Danger Button

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Danger" className="myBtn" actionType="danger">Danger</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button title="Danger" className="myBtn" actionType="danger" disabled>Danger Disabled</Button>
  </div>
</div>
```

#### Full Width Button

To make any button take up the full width of its container, add `fullWidth` prop.

```jsx
const divStyles = {
  borderColor: "#888888",
  borderStyle: "solid",
  borderWidth: 1,
  display: "flex",
  flexDirection: "column",
  height: 100,
  justifyContent: "space-around",
  marginLeft: "auto",
  marginRight: "auto",
  padding: 10,
  width: 500
};

<div style={divStyles}>
  <Button className="myBtn" fullWidth>Full Width Default</Button>
  <Button className="myBtn" actionType="danger" fullWidth>Full Width Danger</Button>
</div>
```
