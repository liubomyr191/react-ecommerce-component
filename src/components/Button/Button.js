import React, { Component } from "react";
import PropTypes from "prop-types";
import preventAccidentalDoubleClick from "../../helpers/preventAccidentalDoubleClick";
import { containerStyles, fullWidthContainerStyles, fullWidthTextContainerStyles } from "./styles";

class Button extends Component {
  static propTypes = {
    /**
     * The type of action performed by the button
     */
    actionType: PropTypes.oneOf(["danger", "default", "important", "secondary"]),
    /**
     * The contents of the button, such as text, icons, or any combination of React and HTML components
     */
    children: PropTypes.node,
    /**
     * Classes for the button element
     */
    className: PropTypes.string,
    /**
     * Set to `true` to prevent the button from calling `onClick` when clicked
     */
    disabled: PropTypes.bool,
    /**
     * Button should take full width
     */
    fullWidth: PropTypes.bool,
    /**
     * Called with no arguments whenever the button is clicked. There is double-click protection,
     * so if the user double-clicks quickly, onClick is called only once.
     */
    onClick: PropTypes.func,
    /**
     * The string for the `title` attribute of the button element
     */
    title: PropTypes.string
  };

  static defaultProps = {
    actionType: "default",
    disabled: false,
    fullWidth: false,
    onClick() {}
  };

  handleClick = preventAccidentalDoubleClick((event) => {
    event.preventDefault();

    const { disabled, onClick } = this.props;

    if (!disabled) {
      onClick();
    }
  });

  handleKeyPress = (event) => {
    if (event.keyCode === 13) this.handleClick(event);
  };

  render() {
    const { children, className, disabled, fullWidth, title, actionType } = this.props;

    // Build className
    const finalClassName = `rui--btn rui--btn-${actionType} ${disabled ? "rui--btn-disabled " : ""}${className}`.trim();

    // Build style
    let style;
    if (fullWidth) {
      if (typeof children === "string") {
        style = fullWidthTextContainerStyles;
      } else {
        style = fullWidthContainerStyles;
      }
    } else {
      style = containerStyles;
    }

    return (
      <div
        className={finalClassName}
        onClick={this.handleClick}
        onKeyPress={this.handleKeyPress}
        role="button"
        style={style}
        tabIndex={0}
        title={title}
      >
        {children}
      </div>
    );
  }
}

export default Button;
