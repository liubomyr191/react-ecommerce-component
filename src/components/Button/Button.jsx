import React, { Component } from "react";
import PropTypes from "prop-types";
import preventAccidentalDoubleClick from "../../helpers/preventAccidentalDoubleClick";
import { containerStyles, flexContainerStyles } from "./styles";

class Button extends Component {
  static propTypes = {
    /**
     * The contents of the button, such as text, icons, or any combination of React and HTML components
     */
    children: PropTypes.node,
    /**
     * Classes for the button element
     */
    className: PropTypes.string,
    /**
     * Set to `true` to prevent the button from calling `onClick` when clicked
     */
    disabled: PropTypes.bool,
    /**
     * Called with no arguments whenever the button is clicked. There is double-click protection,
     * so if the user double-clicks quickly, onClick is called only once.
     */
    onClick: PropTypes.func,
    /**
     * The string for the `title` attribute of the button element
     */
    title: PropTypes.string
  };

  static defaultProps = {
    disabled: false,
    onClick() {}
  };

  onClick = preventAccidentalDoubleClick((event) => {
    const { disabled, onClick } = this.props;

    if (!disabled) {
      onClick(event);
    }
  });

  render() {
    const { children, className, disabled, title } = this.props;

    const finalClassName = `btn-default ${disabled ? "btn-disabled " : ""}${className}`.trim();
    const style = typeof children === "string" ? containerStyles : flexContainerStyles;

    return (
      <div className={finalClassName} onClick={this.onClick} style={style} title={title}>
        {children}
      </div>
    );
  }
}

export default Button;
