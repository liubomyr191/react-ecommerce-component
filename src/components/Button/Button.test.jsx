import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Button from "./Button";

test("basic snapshot", () => {
  const component = renderer.create(<Button title="title" className="a b">Submit</Button>);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("calls onClick when clicked", () => {
  const clickSpy = jest.fn();
  const item = shallow(<Button onClick={clickSpy} />);

  item.simulate("click");

  expect(clickSpy).toHaveBeenCalled();
});

it("calls onClick only once when double-clicked before timeout", () => {
  const clickSpy = jest.fn();
  const item = shallow(<Button onClick={clickSpy} />);

  item.simulate("click");
  item.simulate("click");

  expect(clickSpy.mock.calls.length).toBe(1);
});

it("calls onClick twice when double-clicked after timeout", (done) => {
  const clickSpy = jest.fn();
  const item = shallow(<Button onClick={clickSpy} />);

  item.simulate("click");
  // The debounce timeout for double clicking the button is 600ms
  // so we must wait until that is over before clicking it again.
  setTimeout(() => {
    item.simulate("click");
    expect(clickSpy.mock.calls.length).toBe(2);
    done();
  }, 750);
});
