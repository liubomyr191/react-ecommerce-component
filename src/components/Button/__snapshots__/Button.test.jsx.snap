// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<div
  className="a b"
  onClick={[Function]}
  title="title"
>
  Submit
</div>
`;
